src/runtime/runtime.ts

```diff
@@ -459,84 +460,276 @@ interface Constraint {
+    let resolved = this.resolve(prefix);
+
+    // If none of our returns are unbound
+    // @NOTE: We don't need to worry about the filter case here, since he'll be
+    let unresolvedOutput = false;
+    for(let output of this.returnNames) {
+      if(resolved[output] === undefined) {
+        unresolvedOutput = true;
+        let field = this.fields[output];
+        if(isRegister(field)) {
+          proposal.forRegisters.push(field);
+        }
+      }
+    }
+    if(!unresolvedOutput) {
+      proposal.skip = true;
+      return proposal;
+    }
+
+    // If any of our args aren't resolved yet, we can't compute results either.
+    // @NOTE: This'll need to be touched up when we add optional support if they
+    // co-inhabit the args object.
+    for(let input of this.argNames) {
+      if(resolved[input] === undefined) {
+        proposal.skip = true;
+        return proposal;
+      }
+    }
+
+    // Otherwise, we're ready to propose.
+    proposal.skip = false;
+
+    if(this.estimate) {
+      // If the function provides a cardinality estimator, invoke that.
+      proposal.cardinality = this.estimate(index, prefix, transaction, round);
+
+    } else {
+      // Otherwise, we'll just return 1 for now, since computing a function is almost always cheaper than a scan.
+      // @NOTE: If this is an issue, we can just behave like scans and compute ourselves here, caching the results.
+      proposal.cardinality = 1;
+    }
+
+    return proposal;
+  }
+
+  /** Pack the resolved register values for the functions argument fields into an array. */
+  packInputs(resolved:ResolvedFields) {
+    let inputs = this.applyInputs;
+    let ix = 0;
+    for(let argName of this.argNames) {
+      // If we're asked to resolve the propoal we know that we've proposed, and we'll only propose if these are resolved.
+      inputs[ix] = GlobalInterner.reverse(resolved[argName]!);
+      ix++;
+    }
+    return inputs;
+  }
+
+  unpackOutputs(outputs:undefined|RawValue[]) {
+    if(!outputs) return;
+    for(let ix = 0; ix < outputs.length; ix++) {
+      outputs[ix] = GlobalInterner.intern(outputs[ix]);
+    }
+    return outputs as ID[];
   }
 
-  resolveProposal(index:Index, proposal:Proposal, transaction:number, round:number, results:any[]):ID[][] {
+  resolveProposal(index:Index, prefix:ID[], proposal:Proposal, transaction:number, round:number, results:any[]):ID[][] {
+    let resolved = this.resolve(prefix);
+
+    // First we build the args array to provide the apply function.
+    let inputs = this.packInputs(resolved);
+
+    // Then we actually apply it and then unpack the outputs.
+    // @FIXME: We don't have any intelligent support for not computing unnecessary returns atm.
+    // @FIXME: We only support single-return atm.
+    let outputs = this.unpackOutputs(this.apply.apply(this, inputs));
+    if(!outputs) return results;
+
+    // Finally, if we had results, we create the result prefixes and pass them along.
```

Comment:
Make sure the bound return registers line up with our outputs too, rather than just ignoring them.